### PR TITLE
Impl forwarder contract

### PR DESCRIPTION
**Pull request by nxqbao**
### Description
A vault contract that keeps RON, and behaves as an EOA account to interact with a target contract.

There are three roles of interaction:
- Admin: top-up and withdraw RON to the vault, cannot forward call to the target.
- Moderator: forward all calls to the target, can top-up RON, cannot withdraw RON.
- Others: can top-up RON, cannot execute any other actions.

### Implementation
Inspired by [OpenZeppelin TransparentUpgradeableProxy](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/transparent/TransparentUpgradeableProxy.sol)

### Checklist
- [x] I have clearly commented on all the main functions following the [NatSpec Format](https://docs.soliditylang.org/en/v0.8.0/natspec-format.html)
- [x] The box that allows repo maintainers to update this PR is checked
- [x] I tested locally to make sure this feature/fix works

### CODE REVIEW

**Comment by nxqbao** [comment]
Bug: unauthorized account can still call external functions exposed by `Forwarder` contract.

**Review by ducthotran2010 — COMMENTED** [review]
Please follow the layout guidelines: https://docs.soliditylang.org/en/v0.8.17/style-guide.html#order-of-layout

File: contracts/extensions/forwarder/ForwarderStorage.sol [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  function _getAdmin() internal view returns (address) {
+    return StorageSlot.getAddressSlot(_ADMIN_SLOT).value;
+  }
+
+  /**
+   * @dev Stores a new address in the EIP1967 admin slot.
+   */
+  function _setAdmin(address newAdmin) private {
+    require(newAdmin != address(0), "ERC1967: new admin is the zero address");
```
[the remark this answers is not in the document]
Comment (nxqbao) [reply contracts/extensions/forwarder/ForwarderStorage.sol:68]:
```suggestion
    require(newAdmin != address(0), "ForwarderStorage: new admin is the zero address");
```

File: contracts/ronin/VaultForwarder.sol [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * @title A vault contract that keeps RON, and behaves as an EOA account to interact with a target contract.
+ * @dev There are three roles of interaction:
+ * - Admin: top-up and withdraw RON to the vault, cannot forward call to the target.
+ * - Moderator: forward all calls to the target, can top-up RON, cannot withdraw RON.
+ * - Others: can top-up RON, cannot execute any other actions.
+ */
+contract VaultForwarder is Forwarder, RONTransferHelper {
+  event ForwarderWithdrawn(address indexed _recipient, uint256 _value);
```
[the remark this answers is not in the document]
Comment (nxqbao) [reply contracts/ronin/VaultForwarder.sol:16]:
```suggestion
  /// @dev Emitted when the admin withdraws all RON from the forwarder contract.
  event ForwarderRONWithdrawn(address indexed _recipient, uint256 _value);
```

File: contracts/extensions/forwarder/Forwarder.sol [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "./ForwarderLogic.sol";
+import "./ForwarderStorage.sol";
+
+contract Forwarder is ForwarderLogic, ForwarderStorage, AccessControl {
+  /// @dev Moderator of the forwarder role hash
+  bytes32 public constant MODERATOR_ROLE = keccak256("MODERATOR_ROLE");
+
+  /**
+   * @dev Initializes the forwarder with an initial target specified by `__target`.
```
[the remark this answers is not in the document]
Comment (ducthotran2010) [reply contracts/extensions/forwarder/Forwarder.sol:10]:
```suggestion
   * @dev Initializes the contract storage.
```

Or

```suggestion
   * @dev Initializes the forwarder with an initial target address and a contract admin.
```

File: contracts/extensions/forwarder/ForwarderRole.sol [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  /**
+   * @dev Returns the current target address.
+   */
+  function _getTarget() internal view returns (address) {
+    return StorageSlot.getAddressSlot(_TARGET_SLOT).value;
+  }
+
+  /**
+   * @dev Stores a new address in the EIP1967 target slot.
```
Comment (ducthotran2010) [inline contracts/extensions/forwarder/ForwarderRole.sol:58]:
```suggestion
   * @dev Stores a new address in the target slot.
```

File: contracts/extensions/forwarder/ForwarderRole.sol [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  /**
+   * @dev Returns the current admin address.
+   */
+  function _getAdmin() internal view returns (address) {
+    return StorageSlot.getAddressSlot(_ADMIN_SLOT).value;
+  }
+
+  /**
+   * @dev Stores a new address in the EIP1967 admin slot.
```
Comment (ducthotran2010) [inline contracts/extensions/forwarder/ForwarderRole.sol:30]:
```suggestion
   * @dev Stores a new address in the admin slot.
```

File: contracts/extensions/forwarder/ForwarderLogic.sol [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  ) internal {
+    (bool _success, bytes memory _res) = __target.call{ value: __value }(__data);
+
+    if (!_success) {
+      require(_res.length >= 4, "Forwarder: target reverts silently");
+      assembly {
+        let ptr := mload(0x40)
+        let size := returndatasize()
+        returndatacopy(ptr, 0, size)
+        revert(ptr, size)
```
Comment (ducthotran2010) [inline contracts/extensions/forwarder/ForwarderLogic.sol:25]: For gas saving:
```suggestion
        let size := returndatasize()
        revert(_res, size)
```

File: contracts/extensions/forwarder/Forwarder.sol [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  /**
+   * @dev Forwards the encoded call specified by `_data` to the target. The forwarder attachs `_val` value
+   * from the forwarder contract and sends along with the call.
+   *
+   * Requirements:
+   * - Only moderator can call this method.
+   */
+  function functionCall(bytes memory _data, uint256 _val) external payable onlyModerator {
```
Comment (ducthotran2010) [inline contracts/extensions/forwarder/Forwarder.sol:125]: I guess the contract admin has some uses case to call this

**Review by ducthotran2010 — APPROVED** [review]

**Pull request merged**